### PR TITLE
Refine validation

### PR DESCRIPTION
**Pull request by lukpueh**
**Fixes issue #**:
Post-merge refinement of @adityasaky's #25.

**Description of pull request**:
Simplifies json required field validation, adds artifact rule validation function and some more tests (with TDT setup).

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by lukpueh** [comment]
IMO the validation error messages could be refined some more, to always show what object is invalid, which part of it, and what's the correct value. We can revisit with  #4.

File: in_toto/model.go [hunk 1]
```diff
@@ -211,6 +222,16 @@ func validateSupplyChainItem(supplyChainItem SupplyChainItem) error {
 	if supplyChainItem.Name == "" {
 		return fmt.Errorf("name cannot be empty")
 	}
+	for _, rule := range supplyChainItem.ExpectedMaterials {
```
Comment (adityasaky) [inline in_toto/model.go]: Previously, we discussed having a helper for iterating and verifying signatures. We also do have something similar in paramsub, though that was messier.

https://github.com/in-toto/in-toto-golang/blob/0e8eb5b8845baffc91cd4b494772b2b77c4ddf0f/in_toto/model.go#L98-L120

https://github.com/in-toto/in-toto-golang/blob/7e0a1a42ec0c1d5d894bd576fbd8fed1cb01da12/in_toto/verifylib.go#L637-L655

Think that's relevant here?

File: in_toto/model.go [hunk 2]
```diff
@@ -202,6 +202,17 @@ type SupplyChainItem struct {
...
+validateArtifactRule calls UnpackRule to validate that the passed rule conforms
+with any of the available rule formats.
+*/
+func validateArtifactRule(rule []string) error {
+	if _, err := UnpackRule(rule); err != nil {
+		return err
+	}
+	return nil
+}
+
```
Comment (adityasaky) [inline in_toto/model.go]: Should there be an explicit `TestValidateArtifactRule` for this?

Comment (adityasaky) [inline in_toto/model.go]: Should there be an explicit `TestValidateArtifactRule` for this?

File: in_toto/model.go (hunk 1 above)
Comment (lukpueh) [reply in_toto/model.go]: Not sure how it is relevant. Can you elaborate?

File: in_toto/model.go (hunk 2 above)
Comment (lukpueh) [reply in_toto/model.go]: IMO we're good here. It's only a wrapper for `UnpackRule`, which is extensively tested in `rulelib_test.go`. And it's covered by higher-level validate functions, such as `TestValidateSupplyChainItem`.

File: in_toto/model.go (hunk 1 above)
Comment (adityasaky) [reply in_toto/model.go]: Basically, I’m asking if there should be a verifyArtifactRules that in turn iterates and calls verifyArtifactRule, similar to how it’s setup for signatures etc elsewhere.

Comment (lukpueh) [reply in_toto/model.go]: Done in 0766aa7

**Comment by lukpueh** [comment]
Thanks for your review, @adityasaky. Let me know if it's good to merge. :)

**Comment by adityasaky** [comment]
Looks good to me.

**Pull request merged**